File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/GetReceivedFriendRequestsCount.java
package uk.co.victoriajanedavis.chatapp.domain.interactors;

import android.util.Log;
import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.RetrieveInteractor;

import javax.inject.Inject;
import java.util.List;


public class GetReceivedFriendRequestsCount implements RetrieveInteractor<Void, Integer> {

    @NonNull
    private final GetReceivedFriendRequestsList requestsList;

    @Inject
    public GetReceivedFriendRequestsCount(@NonNull final GetReceivedFriendRequestsList requestsList) {
        this.requestsList = requestsList;
    }

    @NonNull
    @Override
    public Observable<Integer> getBehaviorStream(@NonNull final Void params) {
        return requestsList.getBehaviorStream(params)
                .map(List::size);
    }
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/GetReceivedFriendRequestsCount.kt
package uk.co.victoriajanedavis.chatapp.domain.interactors

import io.reactivex.Observable
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.RetrieveInteractor
import uk.co.victoriajanedavis.chatapp.domain.common.StreamState
import uk.co.victoriajanedavis.chatapp.domain.common.StreamState.*

import javax.inject.Inject


class GetReceivedFriendRequestsCount @Inject constructor(
    private val requestsList: GetReceivedFriendRequestsList
) : RetrieveInteractor<Void, StreamState<Int>> {

    override fun getBehaviorStream(params: Void?): Observable<StreamState<Int>> {
        return requestsList.getBehaviorStream(null)
            .map { streamState -> when(streamState) {
                is OnNext -> OnNext(streamState.content.size)
                is OnError -> OnError(streamState.throwable)
            }}
    }
}
